arrows/klv/klv_parse.cxx

```diff
@@ -2,433 +2,479 @@
 
 namespace {
 
+// ---------------------------------------------------------------------------
+// Reads a big-endian-encoded unsigned integer from a sequence of bytes
+template < class T, class Iterator >
+T
+read_uint_msb( Iterator data_begin, Iterator data_end )
+{
+  // Ensure types are compatible with our assumptions
+  static_assert( std::is_integral<T>::value, "T must be an integer type" );
+  static_assert( std::is_unsigned<T>::value, "T must be an unsigned type" );
+  static_assert(
+    std::is_same< typename std::decay< decltype( *data_begin ) >::type,
+                  uint8_t >::value, "Iterator must point to uint8_t" );
+
+  // Ensure iterators are compatible with our assumptions
+  ptrdiff_t const distance = std::distance( data_begin, data_end );
+  if ( distance < 0 )
+  {
+    VITAL_THROW( kv::invalid_value, "end iterator cannot be before begin" );
+  }
+
+  auto const length = static_cast< size_t >( distance );
+  if ( sizeof( T ) <= length )
+  {
+    VITAL_THROW( kv::invalid_value, std::string() +
+                 "specified type too small too hold data (" +
+                 std::to_string( sizeof(T) ) + " < " +
+                 std::to_string(length) + ")" );
+  }
+
+  // Functor to insert each successive byte into the output value
+  auto const accumulator = [ ]( T value, uint8_t byte ) {
+                             return ( value << 8 ) | byte;
+                           };
+
+  // Reduce span to final result
+  return std::accumulate( data_begin, data_end, static_cast< T >( 0 ),
+                          accumulator );
+}
+
+// ---------------------------------------------------------------------------
+// If unprintable chars detected, replaces them with '.' and appends hex of
+// entire string
 std::string
-FormatString( std::string const& val )
+format_string( std::string const& value )
 {
-  const char hex_chars[16] = { '0', '1', '2', '3', '4', '5', '6', '7',
-                               '8', '9', 'A', 'B', 'C', 'D', 'E', 'F' };
-  const size_t len( val.size() );
-  bool unprintable_found(false);
-  std::string ascii;
-  std::string hex;
-
-  for (size_t i = 0; i < len; i++)
+  // Prints value, but with any unprintable chars replaced with '.'
+  bool print_hex = false;
+  std::stringstream ss;
+  for( char c : value )
   {
-    char const byte = val[i];
-    if ( ! isprint( byte ) )
+    // See https://en.cppreference.com/w/cpp/string/byte/isprint for rationale
+    // for static_cast
+    if( std::isprint( static_cast< unsigned char >( c ) ) )
     {
-      ascii.append( 1, '.' );
-      unprintable_found = true;
+      ss << c;
     }
     else
     {
-      ascii.append( 1, byte );
+      ss << '.';
+      print_hex = true;
     }
+  }
 
-    // format as hex
-    if (i > 0)
+  // Outputs like " (01 5A 2C 00)"
+  if( print_hex )
+  {
+    ss << " (";
+    ss << std::hex << std::uppercase << std::setfill( '0' ) << std::setw( 2 );
+    for( auto it = value.cbegin(); it != value.cend(); ++it )
```

Comment:
Is there a way to do the test on line 104 with a range-based for?

Reply:
```c++
auto first = true;
for( auto c : value )
{
  first ? first = false : ( ss  << ' ', false );
  ss << static_cast< unsigned int >( c );
}
```

(Or write out the `if` instead of abusing the ternary operator 🙂. Something along these lines, i.e. having a 'first flag', seems fairly common, at any rate.)